XYPanel: Improvements

mdvictor opened the pull request:
- [x] Show `prepScatter` errors to user
- [x] Fix explicit mode to support multiple 'layers' (see geomap) 
- [x] Improve tooltip and apply tooltip options
- [x] Improve legend and apply legend options
- [ ] Weird lines appear on refresh in some cases (gdev dashboard shows this happening)
- [x] rename XY - Explicit to Auto - Manual
- [x] Enable data links
- [x] Migrated PureComponent to FunctionalComponent

More points TBD

mdvictor committed:
Fix panel option bugs and make tooltip options work

codeincarnate commented:
Part of #54192

mdvictor commented:
TODO: rename XY - Explicit to Auto - Manual

mdvictor committed:
Support multiple series in explicit mode. Rename XY/Explicit to Auto/Manual

mdvictor committed:
Fixes

mdvictor committed:
Fix

mdvictor commented:
test data links and annotations

mdvictor committed:
Legend improvements

mdvictor committed:
Rewrite Pure Component to Function Component

mdvictor committed:
Add datalinks support

codeincarnate commented:
Nitpicking but it  looks like the label background is off here:

https://user-images.githubusercontent.com/199847/188040862-adcf8839-4635-42ce-9044-e2c2c2370872.mov

codeincarnate commented:
Looks like series selection with the legend isn't working 😢 

https://user-images.githubusercontent.com/199847/188041552-0cbe6541-3b82-4fa4-8739-9e65214856ec.mov

codeincarnate commented on public/app/plugins/panel/xychart/XYChartPanel2.tsx:
```diff
@@ -13,116 +22,225 @@ import {
-  componentDidMount() {
-    this.initSeries(); // also data
-  }
+  const isToolTipOpen = useRef<boolean>(false);
```
This needs to be a ref?

mdvictor replied:
I tried with state but it didn't work out. Not entirely sure why but I'm going to leave it a ref for now and look over it later.

codeincarnate commented on public/app/plugins/panel/xychart/XYChartPanel2.tsx:
```diff
@@ -13,116 +22,225 @@ import {
+    <>
+      <VizLayout width={props.width} height={props.height} legend={renderLegend()}>
+        {(vizWidth: number, vizHeight: number) => (
+          // <pre style={{ width: vizWidth, height: vizHeight, border: '1px solid green', margin: '0px' }}>
```
I think this can be killed 😄

codeincarnate commented on public/app/plugins/panel/xychart/config.ts:
```diff
@@ -34,16 +32,16 @@ export function getScatterFieldConfig(cfg: ScatterFieldConfig): SetFieldConfigOp
           path: 'point',
           name: 'Points',
-          category: categoryStyles,
+          category: undefined,
```
Should this be removed as opposed to undefined?

codeincarnate reviewed (approved):
Epic PR, lots of really great changes. I think we should keep in mind my comments (of course 😆) but I don't see any reason this shouldn't be merged currently.

mdvictor commented on public/app/plugins/panel/xychart/ManualEditor.tsx:
```diff
@@ -0,0 +1,195 @@
+      <Button icon="plus" size="sm" variant="secondary" onClick={createNewSeries}>
+        Add series
+      </Button>
+      <br />
```
Modified the html

mdvictor commented on public/app/plugins/panel/xychart/ManualEditor.tsx:
```diff
@@ -0,0 +1,195 @@
+  const [selected, setSelected] = useState<number>(-1);
+  const style = useStyles(getStyles);
+
+  const onXFieldChange = (x: string | undefined, index: number) => {
```
Indeed. Fixed.

mdvictor commented:
> Looks like series selection with the legend isn't working 😢
> 
>  Screen.Recording.2022-09-01.at.8.35.54.PM.mov

Yep, that was the last thing to be done. Pushing it asap.

mdvictor commented:
> Nitpicking but it looks like the label background is off here:
> 
>  Screen.Recording.2022-09-01.at.8.29.38.PM.mov

fixed

mdvictor committed:
Legend fixes and CR modifications

mdvictor commented:
Fixed legend default position to be on left (like in other panels)
Introduced legend toggle on label click.
Fixed (most) CR comments

There are still some issues to be fixed that I have not yet pinpointed exactly.
Changing datasources sometimes causes errors to be thrown that should be caught and handled.
CSV Content mode in the TestDatasource crashes the browser entirely (something is very wrong with the panel here)

I would like to fix these issues before merging

mdvictor committed:
Fix bugs that crash panel

mdvictor committed:
Merge branch 'main' into victor/xy-panel-improvements
